src/wallet/wallet.cpp

```diff
@@ -3037,7 +3042,8 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend,
                     nIn++;
                 }
 
-                const unsigned int nBytes = ::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION);
+                // account for additional payloads in fee calculation
+                const unsigned int nBytes = ::GetSerializeSize(txNew, SER_NETWORK, PROTOCOL_VERSION) + nExtraSize;
```

Comment:
I'm thinking about this, why this is needed?

In the `FundTransaction` flow, the user is providing the tx in hex format and the node is parsing it. If there would be an extra payload or sapling data inside the tx, this line `GetSerializeSize(tx) + extraSize` will sum them twice as `GetSerializeSize` calls internally to the tx serialization method which is already taking into account the two fields.

Reply:
> In the FundTransaction flow, the user is providing the tx in hex format and the node is parsing it.

Mhm... there is some confusion.
This is the wallet function. There is no hex format, or parsing, directly involved here.
A generic caller to `CWallet::FundTransaction` provides the transaction to be funded, as a CMutableTransaction `tx` (not an hex string).

The outputs of this mutable tx are used to fill `vecSend`, passed to `CWallet::CreateTransaction`.
With the input data, CreateTransaction creates a mutable transaction object `txNew`.

Since, in order to select enough inputs, the transaction fee must be known, the serialized size is calculated inside `CreateTransaction`.
This calculation is obviously done on the newly created CMutableTransaction (the `txNew` object), which has no extra payload accounted for.

Also, while currently `CWallet::FundTransaction` is called exclusively from the RPC command `fundrawtransaction` (after parsing the user provided hex), soon there will be more RPCs (which involve no parsing of raw txes) relying on this wallet function.

> this line GetSerializeSize(tx) + extraSize will sum them twice

No. You can easily verify, trying to remove this on #2267.
The tests will fail due to fee too low.
Also note that `txNew` is being serialized (not the mutable `tx` passed to FundTransaction).